### ISSUE TITLE
gpu_cpu_trace_compare compiles again and fails: GPU L2 diverged 91.92% from CPU on a synthetic fixture (cause undetermined)

### ISSUE THREAD

**Issue by noahgift** [paiml/aprender#3895]
Found by repairing the dark test targets on #3809. **This is a report that a dark test now compiles and fails. It is NOT a report of a GPU defect** — see "What this does not show".

## Measured

At `712f6d433`, with the three struct-literal repairs from `fix/3809-dark-target-repair` (`1096e2969`):

```
cargo test -p aprender-serve --test driver_cuda_gguf        --features cuda   11 passed, 0 failed
cargo test -p aprender-serve --test phase21_trace_divergence --features cuda    4 passed, 0 failed
cargo test -p aprender-serve --test gpu_cpu_trace_compare   --features cuda    0 passed, 1 FAILED

  tests::test_trace_comparison, gpu_cpu_trace_compare.rs:310
  PMAT-216 FAIL: GPU L2 diverged 91.92% from CPU (max 1%)
```

## Why this test had never reported anything

It could not be **built**. `AprTransformer` gained `lm_head_tied` and the test's literal was never updated, so the target failed to compile — and CI runs `--lib` plus an explicit list of `--test` targets that does not include this one. It was not skipped or failing; it was absent from the build graph.

Note also that the failure only appears **with `--features cuda`**: bare `cargo check -p aprender-serve --test gpu_cpu_trace_compare` is `rc=0`, because the bodies are cfg-gated out. Any build-graph guard that does not enumerate feature sets will report green here.

## What this does NOT show

**It is not evidence of a GPU correctness defect in a production path.** The test drives a **synthetic `create_test_model()` fixture**, not a real model. Three candidate causes, in the order they should be eliminated:

1. **The fixture has drifted.** It could not compile for an unknown period, so it was never updated alongside the code it exercises. A fixture that no longer sets up what the traced path requires would produce arbitrary divergence.
2. **A real divergence in the traced path.**
3. **The repair values.** Least likely: all three are the fields' documented defaults (`lm_head_tied: false` for a fixture that materialises a separate `lm_head_weight`), and the two sibling targets pass with the same values.

**The caution is on this release's record.** tinyllama was reported as a CUDA correctness defect blocking a tag, and it failed identically on CPU — the gate had hardcoded "(CPU passed)" about a leg it had not judged (#3870). A bare "92% GPU divergence" is exactly the kind of number that gets quoted before it is understood.

## Suggested first step

Eliminate (1) before investigating (2): confirm the synthetic fixture still constructs a model the traced path considers well-formed, and check whether the divergence reproduces on a real GGUF through the same comparison. The test prints a ratio analysis (`avg`, `std`, and "Consistent scale factor detected: CPU ≈ Nx GPU") immediately before the assertion — **if that std is small, a consistent scale factor points at a fixture/normalisation mismatch rather than at numerical divergence**, and that is one run away from being known.

## Why it matters for #3809 beyond one test

The epic's thesis is "tests that exist and never execute". This one is worse: it could not be built, **so nobody could learn that it fails.** The rot hid a signal, not just a test — whatever that signal turns out to mean.

**Comment by noahgift** [comment]
## The ratio analysis rules out the benign explanation I suggested, and narrows it

I proposed checking whether the printed ratio std was small, because a consistent scale factor would point at a fixture/normalisation mismatch rather than divergence. **It is not small.** From the same run:

```
=== CPU Forward (traced via TracedForward trait) ===
Embed: mean=0.8667, std=0.0880, L2≈6.9691
CPU argmax: 196

=== GPU Forward (traced via TracedForward trait) ===
GPU Logits: mean=-1.8126, std=0.5123, min=-2.6840, max=-0.9217
GPU argmax: 0

CPU L2: 372.9509, GPU L2: 30.1339, diff: 91.92%
Ratio analysis: avg=-1.52, std=21.5042
```

`std=21.50` against `avg=-1.52` — the ratios are not consistent, so **this is not a uniform scale factor.** That eliminates the cheapest benign explanation.

## What the numbers point at instead

- **`GPU argmax: 0`.** Token id 0 is the signature this repo already has a name for: #3782 records that `"!!!!!!!!"` is token id 0 in the Qwen vocab, *"i.e. exactly what a model with dead logits emits"*.
- **GPU logits span `[-2.68, -0.92]`, mean `-1.81`, std `0.51`** — a narrow, entirely negative band, which is what a logit vector looks like when it carries no signal rather than a rescaled version of the right answer.
- **GPU L2 is 30.1 against the CPU's 372.9**, ~12x smaller in magnitude, with inconsistent per-element ratios.

Taken together this reads as **the GPU path producing degenerate output**, not as a scaled or slightly-drifted version of the CPU result.

## What it still does NOT establish

The fixture is synthetic. Degenerate GPU logits are equally consistent with **the fixture failing to provide what the GPU path needs to upload** (it sets `q4k_layers: None`, and the GPU side may have nothing real to work from) as with a defect in the traced GPU path itself. Those two are distinguishable, and the distinguishing run is the one already suggested: drive the same comparison from a **real GGUF** rather than `create_test_model()`.

So the revision is: not "a possible scale-factor artefact" but **"the GPU leg of this comparison emits dead logits on this fixture; determine whether the fixture can legitimately drive the GPU path at all before concluding anything about the path."**

Still not a GPU correctness claim, and still 0.70.0.

**Comment by noahgift** [comment]
milestone → 0.71.0 per docs/roadmaps/milestone-triage-2026-09-23.md (#4024), operator approval 2026-09-23 (relayed by aprender-cf). Reason: GPU L2 diverges 91.92% from CPU: backend correctness (#3994)
